### PR TITLE
feat: comms integration

### PR DESCRIPTION
**Pull request by nduchak**
Add logic to create event based on the event group

### CODE REVIEW

File: src/sequelize.ts [hunk 1]
```diff
@@ -58,6 +58,26 @@ export function BigNumberStringType (propName: string): Partial<ModelAttributeCo
...
 
+/**
+ * consider that the field will be stored as string separated by '|' symbol in the data base,
+ * so you not be able to use array comparision when querying
+ * ArrayStringType for sequelize models
+ * @param propName
+ * @constructor
+ */
+export function ArrayStringType (propName: string): Partial<ModelAttributeColumnOptions> {
+  const separator = '|'
```
Comment (jurajpiar) [inline src/sequelize.ts:69]: why not comma `,`?

File: src/services/notification/index.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    notificationService.hooks(notificationHooks)
+    app.configure(notificationChannels)
+
+    // Init comms
+    const comms = app.get('comms') as Comms
+    comms.messageHandler = messageHandler(notificationService)
+
+    return {
+      stop: () => undefined
+    }
```
Comment (jurajpiar) [inline src/services/notification/index.ts:46]: maybe you could DRY this is a bit with a const.

File: src/services/notification/notification.model.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+})
+export default class NotificationModel extends Model {
+  @Column({ allowNull: false })
+  type!: string
+
+  @Column({ ...ArrayStringType('accounts'), allowNull: false })
+  accounts!: string[]
+
+  @Column({ type: DataType.JSON, allowNull: false })
+  payload!: CommsPayloads & { code: string }
```
Comment (jurajpiar) [inline src/services/notification/notification.model.ts:17]: maybe extract type in the definitions just to make it more obvious.

File: src/services/storage/handlers/offer.ts [hunk 4]
```diff
@@ -53,7 +54,7 @@ const handlers: { [key: string]: Function } = {
     }
     logger.info(`Updating capacity ${offer.totalCapacity} (ID: ${offer.provider})`)
   },
-  async MessageEmitted (event: EventData, offer: Offer, offerService: OfferService): Promise<void> {
+  async MessageEmitted (event: EventData, offer: Offer, offerService: OfferService, { comms }: { comms: Comms }): Promise<void> {
```
Comment (jurajpiar) [inline src/services/storage/handlers/offer.ts:57]: split onto new lines

**Review by jurajpiar — APPROVED** [review]
Looks alright. Could do with improving readability by splitting long lines and such.

File: src/sequelize.ts (hunk 1 above)
Comment (nduchak) [reply src/sequelize.ts:69]: I think about the symbol which is less used. For example, if you have an array of strings you probably can use comma in one of the elements

File: src/services/notification/index.ts (hunk 2 above)
Comment (nduchak) [reply src/services/notification/index.ts:46]: you mean create a const for stop function like: `const emptyStop = () => undefined`?

File: src/services/notification/notification.model.ts (hunk 3 above)
Comment (nduchak) [reply src/services/notification/notification.model.ts:17]: Good point

File: src/services/storage/handlers/offer.ts (hunk 4 above)
Comment (nduchak) [reply src/services/storage/handlers/offer.ts:57]: Done

File: src/services/notification/index.ts (hunk 2 above)
Comment (jurajpiar) [reply src/services/notification/index.ts:46]: yea it's a small point. it's just that it is used twice in the fine so might as well. but it's up to you of course

Comment (nduchak) [reply src/services/notification/index.ts]: Done, thanks

File: src/communication/handlers.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+ * @param notificationService
+ */
+export function messageHandler (
+  notificationService?: NotificationService
+): (message: CommsMessage<CommsPayloads>) => Promise<void> {
+  return async function (message: CommsMessage<CommsPayloads>): Promise<void> {
+    const agreement = await Agreement.findOne({ where: { agreementReference: message.payload.agreementReference } })
+
+    if (!agreement) {
+      logger.verbose(`Agreement ${message.payload.agreementReference} for message not found`)
```
Comment (AuHau) [inline src/communication/handlers.ts:79]: Shouldn't be this an error?

File: src/services/notification/notification.channels.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import '@feathersjs/transport-commons'
+import { Application, ServiceAddresses } from '../../definitions'
+
+const CHANNEL = 'notifications'
+
+function filterByOwner (app: Application, data: any) {
+  const accounts: string[] = data.accounts
+  return app.channel(CHANNEL)
+    .filter(connection => {
+      const connectedAccount: string = connection.ownerAddress
```
Comment (AuHau) [inline src/services/notification/notification.channels.ts:10]: Hmm we should rename it from `ownerAddress` at some point, but that has to be coordinated with UI... Lets keep it for now.

File: src/services/storage/index.ts [hunk 7]
```diff
@@ -189,6 +195,7 @@ const storage: CachedService = {
         confirmationService.removeAllListeners()
         stakingEventsEmitter.stop()
         storageManagerEventsEmitter.stop()
+        commsStop()
```
Comment (AuHau) [inline src/services/storage/index.ts:198]: This should be called on the `app.stop()` level as well.

**Review by AuHau — CHANGES_REQUESTED** [review]
Good job on the splitting, but I am afraid I still don't like that much the way how you initialize and handle libp2p. Please see the comments.

File: config/default.json5 [hunk 8]
```diff
@@ -14,7 +14,20 @@
...
+    }
+  },
+
+  notification: {
+    enabled: true,
+    countOfNotificationPersistedPerAgreement: 10
+  },
+
   blockchain: {
+    networkId: '8545',
```
[the remark this answers is not in the document]
Comment (nduchak) [reply config/default.json5:30]: Done

File: src/sequelize.ts [hunk 9]
```diff
@@ -58,6 +58,26 @@ export function BigNumberStringType (propName: string): Partial<ModelAttributeCo
   }
 }
 
+/**
+ * consider that the field will be stored as string separated by '|' symbol in the data base,
```
[the remark this answers is not in the document]
Comment (nduchak) [reply src/sequelize.ts:62]: Done

File: src/sequelize.ts [hunk 10]
```diff
@@ -58,6 +58,26 @@ export function BigNumberStringType (propName: string): Partial<ModelAttributeCo
...
+ */
+export function ArrayStringType (propName: string): Partial<ModelAttributeColumnOptions> {
+  const separator = '|'
+  return {
+    type: DataType.STRING(),
+    get (this: Model): Array<string> {
+      return this.getDataValue(propName as any).split(separator)
+    },
+    set (this: Model, value: string[]): void {
+      this.setDataValue(propName as any, value.join(separator))
```
[the remark this answers is not in the document]
Comment (nduchak) [reply src/sequelize.ts:76]: Done

File: src/communication/index.ts [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+export function getRoomTopic (offerId: string): string {
+  return `${config.get<string>('blockchain.networkId')}:${offerId}`
+}
+
+export function getRoom (topic: string): Room | undefined {
+  return rooms.get(topic)
+}
+
+export function getPeerId (): string {
```
[the remark this answers is not in the document]
Comment (nduchak) [reply src/communication/index.ts:32]: Done

File: src/communication/handlers.ts (hunk 5 above)
Comment (nduchak) [reply src/communication/handlers.ts:79]: Done

File: src/services/storage/index.ts [hunk 12]
```diff
@@ -162,11 +163,16 @@ const storage: CachedService = {
     }
 
     const eth = app.get('eth') as Eth
+
+    // Init comms
+    await initComms()
```
[the remark this answers is not in the document]
Comment (nduchak) [reply src/services/storage/index.ts:168]: Done

File: src/communication/index.ts [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+} from '../definitions'
+import { errorHandler } from '../utils'
+import { messageHandler } from './handlers'
+
+const logger = loggingFactory('communication')
+
+// (offerId -> room) MAP
+export const rooms = new Map<string, Room>()
+
+let libp2p: Libp2p
```
[the remark this answers is not in the document]
Comment (nduchak) [reply src/communication/index.ts:21]: Then you should refactor pinner as you store it there in the same way

File: src/services/storage/index.ts (hunk 7 above)
Comment (nduchak) [reply src/services/storage/index.ts:198]: Done

File: src/communication/index.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (nduchak) [reply src/communication/index.ts:21]: Done

File: src/communication/index.ts [hunk 14]
```diff
@@ -83,29 +66,33 @@ export async function initLibp2p (): Promise<Libp2p> {
...
-  libp2p = await initLibp2p()
+  app.set('libp2p', await initLibp2p())
+  const notificationService = app.service(ServiceAddresses.NOTIFICATION)
+  _messageHandler = messageHandler(notificationService)
 }
 
-export async function stop (): Promise<void> {
+export async function stop (app: Application): Promise<void> {
   for (const [, room] of rooms) {
     room.leave()
```
Comment (AuHau) [inline src/communication/index.ts:86]: I guess you should remove the room from the `Map` as well right? Just to properly cleanup...

**Review by AuHau — CHANGES_REQUESTED** [review]
Almost there!

File: src/communication/index.ts (hunk 14 above)
Comment (nduchak) [reply src/communication/index.ts]: Yes, Done

**Review by AuHau — APPROVED** [review]
Sweet! Now just resolve the conflicts and we are good to merge!

**Comment by julianmrodri** [comment]
Awesome job lets get this in! Congrats Naz!

**Review by julianmrodri — APPROVED** [review]
LGTM! Awesome work.

**Pull request merged**